src/main/java/fi/aalto/cs/apluscourses/intellij/actions/SubmitExerciseAction.java

```diff
@@ -90,33 +91,61 @@ public void actionPerformed(@NotNull AnActionEvent e) {
       return;
     }
 
-    SubmissionHistory submissionHistory = tryGetSubmissionHistory(
-        exercise, authentication, project);
-    if (submissionHistory == null) {
+    List<Group> groups = tryGetUserGroups(course, authentication, project);
+    if (groups == null) {
       return;
     }
 
-    List<Group> groups = tryGetUserGroups(course, authentication, project);
-    if (groups == null) {
+    Module[] modules = Optional.ofNullable(project)
+        .map(ModuleManager::getInstance)
+        .map(ModuleManager::getModules)
+        .orElseGet(() -> new Module[0]);
+
+    ModuleSelectionViewModel moduleSelectionViewModel = new ModuleSelectionViewModel(modules);
+
+    if (!new ModuleSelectionDialog(project, moduleSelectionViewModel).showAndGet()) {
       return;
     }
 
-    SubmissionViewModel viewModel = new SubmissionViewModel(
-        exercise, submissionHistory, groups, authentication, project);
+    Module selectedModule = moduleSelectionViewModel.selectedModule.get();
 
-    if (!new ModuleSelectionDialog(viewModel).showAndGet()
-        || viewModel.getSelectedModule() == null) {
+    if (selectedModule == null) {
       return;
     }
 
-    List<Path> filePaths = tryGetFilePaths(
-        exercise.getFilenames(), viewModel.getSelectedModule(), project);
-    if (filePaths == null) {
+
+    Path basePath = Paths.get(ModuleUtilCore.getModuleDirPath(selectedModule));
+
+    Path[] filePaths;
+    try {
+      filePaths = exercise.getFilePaths(basePath);
+    } catch (FileDoesNotExistException ex) {
+      notifier.notify(new MissingFileNotification(selectedModule.getName(), ex.getName()), project);
       return;
     }
 
-    if (new SubmissionDialog(viewModel).showAndGet()) {
-      // Do actual submission
+    IOException ioException = null;
+
+    while (true) {
+      SubmissionHistory submissionHistory = tryGetSubmissionHistory(
+          exercise, authentication, project);
+      if (submissionHistory == null) {
+        return;
+      }
+
+      SubmissionViewModel viewModel = new SubmissionViewModel(exercise, submissionHistory, groups,
+          authentication, filePaths, ioException);
+
+      if (!new SubmissionDialog(viewModel, project).showAndGet()) {
+        return;
+      }
+
+      try {
+        viewModel.buildSubmission().submit();
+        return;
+      } catch (IOException ex) {
+        ioException = ex;
+      }
```

Comment:
I almost like understand this, but please, consider simplifying it! :nerd_face: 

and a question, what would happen, if (L144) `viewModel.buildSubmission().submit();` keeps throwing an `IOException` :grey_question: